include/envoy/stats/stats.h

```diff
@@ -171,6 +198,60 @@ class Histogram : public virtual Metric {
 
 typedef std::shared_ptr<Histogram> HistogramSharedPtr;
 
+/**
+ * A histogram that is stored in TLS and used to record values per thread. This holds two
+ * histograms, one to collect the values and other as backup that is used for merge process. The
+ * swap happens during the merge process.
+ */
+class ThreadLocalHistogram : public virtual Histogram {
+public:
+  virtual ~ThreadLocalHistogram() {}
+
+  /**
+   * Called in the beginning of merge process. Swaps the histogram used for collection so that we do
+   * not have to lock the histogram in high throughput TLS writes.
+   */
+  virtual void beginMerge() PURE;
+};
+
+class ThreadLocalHistogramImpl;
```

Comment:
I had to forward declare this because with out this, I have to explicitly add `merge(histogram_t* target)` to ThreadLocalHistogram interface which means we have to expose the `libcircl` to stats.h interface definition. Otherwise I have to use `dynamic_cast` in the `ParentHistogramImpl` merge method. LMK if there are any other better solutions.

Reply:
ok. Thinking further, I think agree with you that ThreadLocalHistogram should not be in stats.h as it is not used by others in the system. I moved it to thread_local_store.h. Also I removed the addTlsHistogram method in ParentHistogram as that is an implementation detail. Now I think it looks clean. PTAL.